tianshou/policy/imitation/bcq.py

```diff
@@ -0,0 +1,213 @@
+        each action to get the target Q. Default to 10.
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.policy.BasePolicy` for more detailed
+        explanation.
+    """
+
+    def __init__(
+        self,
+        actor: torch.nn.Module,
+        actor_optim: torch.optim.Optimizer,
+        critic1: torch.nn.Module,
+        critic1_optim: torch.optim.Optimizer,
+        critic2: torch.nn.Module,
+        critic2_optim: torch.optim.Optimizer,
+        vae: VAE,
+        vae_optim: torch.optim.Optimizer,
+        device: Union[str, torch.device] = "cpu",
+        gamma: float = 0.99,
+        tau: float = 0.005,
+        lmbda: float = 0.75,
+        forward_sampled_times: int = 100,
+        num_sampled_action: int = 10,
+        **kwargs: Any
+    ) -> None:
+        # actor is Perturbation!
+        super().__init__(**kwargs)
+        self.actor = actor
+        self.actor_target = copy.deepcopy(self.actor)
+        self.actor_optim = actor_optim
+
+        self.critic1 = critic1
+        self.critic1_target = copy.deepcopy(self.critic1)
+        self.critic1_optim = critic1_optim
+
+        self.critic2 = critic2
+        self.critic2_target = copy.deepcopy(self.critic2)
+        self.critic2_optim = critic2_optim
+
+        self.vae = vae
+        self.vae_optim = vae_optim
+
+        self.gamma = gamma
+        self.tau = tau
+        self.lmbda = lmbda
+        self.device = device
+        self.forward_sampled_times = forward_sampled_times
+        self.num_sampled_action = num_sampled_action
+
+    def train(self, mode: bool = True) -> "BCQPolicy":
+        """Set the module in training mode, except for the target network."""
+        self.training = mode
+        self.actor.train(mode)
+        self.critic1.train(mode)
+        self.critic2.train(mode)
+        return self
+
+    def forward(
+        self,
+        batch: Batch,
+        state: Optional[Union[dict, Batch, np.ndarray]] = None,
+        **kwargs: Any,
+    ) -> Batch:
+        """Compute action over the given batch data."""
+        # There is "obs" in the Batch
+        # obs_group: several groups. Each group has a state.
+        obs_group: torch.Tensor = to_torch(  # type: ignore
+            batch.obs, device=self.device
+        )
+        act = []
+        for obs in obs_group:
+            # now obs is (state_dim)
+            obs = (obs.reshape(1, -1)).repeat(self.forward_sampled_times, 1)
+            # now obs is (forward_sampled_times, state_dim)
+
+            # decode(obs) generates action and actor perturbs it
+            action = self.actor(obs, self.vae.decode(obs))
+            # now action is (forward_sampled_times, action_dim)
+            q1 = self.critic1(obs, action)
+            # q1 is (forward_sampled_times, 1)
+            ind = q1.argmax(0)
+            act.append(action[ind].cpu().data.numpy().flatten())
+        act = np.array(act)
+        return Batch(act=act)
```

Comment:
Can we batch these operations instead of using for-loop?

Reply:
this is because in MLP it uses `flatten(1)` instead of `reshape(-1, dim)`

Reply:
like this:
```python
obs: torch.Tensor = to_torch(
    batch.obs, device=self.device
)
obs = obs.reshape(-1, 1, obs.shape[1]).repeat(1, self.forward_sampled_times, 1)
act = self.actor(obs, self.vae.decode(obs))
q1 = self.critic1(obs, act)
ind = q1.argmax(1)
act = act[range(ind.shape[0]), ind.reshape(-1)].cpu().data.numpy()
```

Reply:
But this requires modification of the MLP and Critic. I think they should fit more dimensional input.

Reply:
will do in the following PR